tests/test_halo_data_transformer.py

```diff
@@ -0,0 +1,452 @@
+    params=[
+        pytest.param((Y_DIM, X_DIM), id="center"),
+        pytest.param((Z_DIM, Y_DIM, X_DIM), id="center_3d"),
+        pytest.param((X_DIM, Y_DIM, Z_DIM), id="center_3d_reverse",),
+        pytest.param((X_DIM, Z_DIM, Y_DIM), id="center_3d_shuffle",),
+        pytest.param((Y_INTERFACE_DIM, X_INTERFACE_DIM), id="interface"),
+        pytest.param(
+            (Z_INTERFACE_DIM, Y_INTERFACE_DIM, X_INTERFACE_DIM,), id="interface_3d",
+        ),
+    ]
+)
+def dims(request, fast):
+    if fast and request.param in (
+        (X_DIM, Y_DIM, Z_DIM),
+        (Z_INTERFACE_DIM, Y_INTERFACE_DIM, X_INTERFACE_DIM,),
+    ):
+        pytest.skip("running in fast mode")
+    return request.param
+
+
+@pytest.fixture
+def shape(nz, ny, nx, dims, n_halos, n_buffer):
+    return_list = []
+    length_dict = {
+        X_DIM: 2 * n_halos + nx + n_buffer,
+        X_INTERFACE_DIM: 2 * n_halos + nx + 1 + n_buffer,
+        Y_DIM: 2 * n_halos + ny + n_buffer,
+        Y_INTERFACE_DIM: 2 * n_halos + ny + 1 + n_buffer,
+        Z_DIM: nz + n_buffer,
+        Z_INTERFACE_DIM: nz + 1 + n_buffer,
+    }
+    for dim in dims:
+        return_list.append(length_dict[dim])
+    return return_list
+
+
+@pytest.fixture
+def extent(n_points, dims, nz, ny, nx):
+    return_list = []
+    extent_dict = {
+        X_DIM: nx,
+        X_INTERFACE_DIM: nx + 1,
+        Y_DIM: ny,
+        Y_INTERFACE_DIM: ny + 1,
+        Z_DIM: nz,
+        Z_INTERFACE_DIM: nz + 1,
+    }
+    for dim in dims:
+        return_list.append(extent_dict[dim])
+    return return_list
+
+
+def _shape_length(shape: Tuple[int]) -> int:
+    """Compute linear size from slices"""
+    length = 1
+    for s in shape:
+        length *= s
+    return length
+
+
+@pytest.fixture
+def quantity(dims, units, origin, extent, shape, numpy, dtype, backend):
+    """A list of quantities whose values are 42.42 in the computational domain and 1
+    outside of it."""
+    sz = _shape_length(shape)
+    print(f"{shape} {sz}")
+    data = numpy.arange(0, sz, dtype=dtype).reshape(shape)
+    if backend == "gt4py_cupy":
+        quantity = Quantity(
+            cp.asnumpy(data),
+            dims=dims,
+            units=units,
+            origin=origin,
+            extent=extent,
+            gt4py_backend="gtcuda",
+        )
+        layout_map = quantity.storage.layout_map
+        print(layout_map)
+    elif backend == "gt4py_numpy":
+        quantity = Quantity(
+            data,
+            dims=dims,
+            units=units,
+            origin=origin,
+            extent=extent,
+            gt4py_backend="gtx86",
+        )
+        layout_map = quantity.storage.layout_map
+        print(layout_map)
+    else:
+        quantity = Quantity(data, dims=dims, units=units, origin=origin, extent=extent)
+    return quantity
+
+
+@pytest.fixture(params=[-0, -1, -2, -3])
+def rotation(request):
+    return request.param
+
+
+def test_data_transformer_allocate(quantity, n_halos):
```

Comment:
Open ended comment. It's a lot harder to understand and test HaloDataTransformer because it's mutable. I imagine you tried to make it immutable and ran into trouble because there are too many arguments that define each of the edge transfers? I can't understand whether these tests cover every edge case, but if you think these tests fully cover the code that's good enough for me.

If you wanted to make HaloDataTransformer immutable, you could make it initializeable from a sequence of an internal "halo edge update" data container object. Even a NamedTuple is fine. Then the "add spec" lines would build a list of e.g. `HaloEdgeUpdateSpec`, and you can create HaloDataTransformer from this list at the end.

Reply:
It looks like you even already defined this data container class (`_ExchangeDataDescription`)

Reply:
Alright, I have changed the way it gets created to insure immutability. That gets rid of the `compile` call which can now be done internally at the end of the `__init__`

Sorry for lagging behind there, does concepts in py are expressed very differently than what I am used too and my lack of structural python coding is showing. Anyway have a look and tell me what you think